src/coreclr/jit/codegencommon.cpp

```diff
@@ -1410,14 +1417,22 @@ bool CodeGen::genCreateAddrMode(
                 tmpMul = mul;
             }
 
+            if (indexIsEffectivelyZero)
+            {
+                mul = 0;
+                rv2 = nullptr;
+            }
             /* Get hold of the array index and see if it's a constant */
-            if (index->IsIntCnsFitsInI32())
+            else if (index->IsIntCnsFitsInI32())
             {
                 /* Get hold of the index value */
                 ssize_t ixv = index->AsIntConCommon()->IconValue();
 
                 /* Scale the index if necessary */
-                ixv *= tmpMul;
+                if (tmpMul)
```

Comment:
So its ok to have `tmpMul == 0` here from the `else` on line 1417 and not scale?

Reply:
Looks like it can be but only only if we also disable jit optimizations and write IL for `array[1 + 2]` because Roslyn will fold cns + cns too again (even in Debug mode)

Reply:
Actually, let me run a few more tests, probably some parts of this logic can be removed

Reply:
@kunalspathak so I've slightly changed the logic - the whole block can be removed in theory but it produced several Tier0 regressions so I decided to keep it + special handled zero index or zero scale, so if `optGetArrayRefScaleAndIndex` returns either 0 or non-zero but `index` represents `0` then we drop the whole index part.

Reply:
What happens for `!(rv2->OperIs(GT_MUL, GT_LSH) && (rv2->gtGetOp2()->IsCnsIntOrI()))` where we set `index = rv2`?

```c++
else
{
    /* May be a simple array. rv2 will points to the actual index */

    index  = rv2;
    tmpMul = mul;
}
```

Reply:
@kunalspathak good point, it explains a single regression in the jit-diffs, but overall the code as is correct, it's just that the block you highlighted handles constant index that is converted to constant offset (this path is never hit in optimized code from what I see)

Reply:
Let me push a quick change

Reply:
@kunalspathak should be correct now and zero diff